pkg/trace/trace.go

```diff
@@ -0,0 +1,217 @@
+package trace
+
+import (
+	"bufio"
+	"crypto/rand"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/valyala/fastrand"
+
+	log "github.com/sirupsen/logrus"
+)
+
+var envs []string = []string{"prod", "dev"}
+var ar_dc []string = []string{"LON", "PAR", "DEN", "PHX", "DFW", "ATL", "BOS", "MHT", "JFK", "CAR", "LHR", "AMS", "FRA", "BOM", "CAL", "DEL", "MAD", "CHE", "FGH", "RTY", "UIO", "MHJ", "HAN", "YHT", "YUL", "MOL", "FOS", "KUN", "SRI", "FGR", "SUN", "PRI", "TAR", "SAR", "ADI", "ERT", "ITR", "DOW", "UQW", "QBF", "POK", "HQZ", "ZAS", "POK", "LIP", "UYQ", "OIK", "TRU", "POL", "NMC", "AZQ"}
+
+func randomHex(n int) string {
+	bytes := make([]byte, n)
+	if _, err := rand.Read(bytes); err != nil {
+		return ""
+	}
+	return hex.EncodeToString(bytes)
+}
+
+func StartTraceGeneration(file string, numTraces int, maxSpans int) {
+	log.Println("Starting generation of ", numTraces, " traces...")
+	traceCounter := 0
+	log.Println("Opening file ", file, "...")
+	log.Println("Opening file trace_service.json...")
+	service_file, err := os.Create("trace_service.json")
+	service_writer := bufio.NewWriter(service_file)
+	if err != nil {
+		log.Fatal(err)
+	}
+	f, err := os.Create(file)
+	w := bufio.NewWriter(f)
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer service_file.Close()
+	defer f.Close()
+	for traceCounter < numTraces {
+		var traceId = randomHex(8)
+		var randomNumberOfSpans = 1 + fastrand.Uint32n(uint32(maxSpans))
+		var prevSpanId = randomHex(8)
+		traceRoot := generateRootTrace(traceId)
+		WriteToFile(traceRoot,w)
+		serviceAndOperationJson := getServiceAndOperationObject(traceRoot["operationName"], traceRoot["process"])
+		WriteToServiceFile(serviceAndOperationJson, service_writer)
+		spanCounter := 0
+		for spanCounter < int(randomNumberOfSpans) {
+			span,spanId := generateChildSpan(traceId, spanCounter,prevSpanId)
+			serviceAndOperationJson := getServiceAndOperationObject(span["operationName"], span["process"])
+			span["traceID"] = traceId
+			WriteToFile(span, w)
+			WriteToServiceFile(serviceAndOperationJson, service_writer)
```

Comment:
do we need to write an entry to the service file each span? or is once per trace enough?

Reply:
The spans inside each trace can have different service IDs so we need to write it at both places